### ISSUE TITLE
Unable to play an hls stream using latest video.js and videojs-contrib-hls.

### ISSUE THREAD

**Issue by sirisian** [videojs/videojs-contrib-hls#653]
The error in question that shows up in Firefox only for Ubuntu 13.04 and Windows 7 (but doesn't show up for Windows 7 Chrome):

```
Specified "type" attribute of "application/x-mpegURL" is not supported. Load of media resource hls/hdmi/index.m3u8 failed.
```

Note that this all works on Windows 10 Edge. Just tested and it's perfect.

I have nginx-rtmp generating an m3u8 file and ts files from a stream. These files work perfectly in ffplay and quicktime. (The URL http://localhost/hls/hdmi/index.m3u8 was used). A sample of the files can be found here:

http://sirisian.com/randomfiles/testfiles.tar.gz

My index.html looks like:

```
<!doctype html>
<html>
<head>
<meta charset="UTF-8">
<title>HLS Test</title>
<link href="video-js.min.css" rel="stylesheet">
<script src="video.min.js"></script>
<script src="videojs-contrib-media-sources.min.js"></script>
<script src="videojs-contrib-hls.min.js"></script>
<script>
window.onload = function()
{
        var player = videojs('video', { /* Options */ }, function()
        {
                console.log('Playing');

                this.play(); // Not needed because of autoplay

                this.on('ended', function()
                {
                        console.log('This should not happen.');
                });
        });
};
</script>
</head>
<body>
<video id="video" controls>
  <source
     src="hls/hdmi/index.m3u8"
     type="application/x-mpegURL">
</video>
</body>
</html>
```

For each min.js and css file I cloned the latest git and then ran npm install and npm run build. For video.js I ran grunt dist as required. I also used the latest 4.4.2 node.js and npm. More verbosely:

> git clone https://github.com/videojs/video.js.git
> cd video.js
> npm install
> grunt dist
> git clone https://github.com/videojs/videojs-contrib-hls.git
> cd videojs-contrib-hls
> npm install
> npm run build
> git clone https://github.com/videojs/videojs-contrib-media-sources.git
> cd videojs-contrib-media-sources
> npm install
> npm run build

I've tested this in Ubuntu 13.04 Firefox 45 and Windows 7 with the latest Chrome and Firefox. Nothing displays for the video. In the network tab I can clearly it successfully requests the index.m3u8 file and then requests 1 ts file then continues to request the index.m3u8 file over and over from then.

(I have already verified that there are no codec issues as I had this partially working with hls.js but it was buggy and this videojs-contrib-hls fixed that bug over a week ago). (The bug in question: https://github.com/videojs/videojs-contrib-hls/issues/483 ). Basically I'm confident that my m3u8 and ts files are correct. I have the program segmenting at 5s intervals. I've tried 1s also with no change and like I said other programs run it fine. It's like video.js can't use or find the videojs-contrib-hls stuff. (I'm not using Flash and my goal is to make this work in HTML5).

**Comment by tenacex** [comment]
@sirisian Any chance you could assemble a test page on a server that causes the error? It would make testing for our team a whole lot easier.

**Comment by sirisian** [comment]
http://sirisian.com/randomfiles/testhls/index.html

Not sure if the timing will be right though playing from an index.m3u8 like that, but that URL contains a full example with all the files I'm using and a snapshot of the index.m3u8 and ts files.

**Comment by tenacex** [comment]
Hmm, when I plug that link into my player, it doesn't load too. I noticed that when I visit: http://sirisian.com/randomfiles/testhls/hls/hdmi/index.m3u8
It doesn't download a playlist file, like it has everywhere else for me. Perhaps you could look into that?

**Comment by sirisian** [comment]
I don't have cross origin enabled for that directory. That's why I linked the tar.gz previously so the files could be ran locally on your test environment.

**Comment by hdezela** [comment]
If you're using the latest versions, you don't need media-sources. Also, like @tenacex said, your m3u8 is being served as document, and not as application/x-mpegURL.

I tried your files with this code and it worked fine:

```
<html>
    <head>
        <meta charset="UTF-8">
        <title></title>
        <script src="video.min.js"></script>
        <script src="videojs-contrib-hls.min.js"></script>
    </head>
    <body>
        <video id=example-video width=960 height=540 class="video-js vjs-default-skin" controls>
            <source src="hls/hdmi/index.m3u8" type="application/x-mpegURL">
        </video>
        <script>
            var player = videojs('example-video');
            player.play();
        </script>
    </body>
</html>
```

**Comment by sirisian** [comment]
Which browser did you use? Also just to be clear. I do not have Flash. (I don't even have it on the device I'll be running it on. If you can test again, but disable the Flash plugin I think you'll see my results). I just tried that exact code with my files and it doesn't play. In Firefox in the console I see:

> GET http://sirisian.com/randomfiles/testhls/hls/hdmi/index.m3u8
> GET http://sirisian.com/randomfiles/testhls/hls/hdmi/1460412184500.ts
> Specified "type" attribute of "application/x-mpegURL" is not supported. Load of media resource hls/hdmi/index.m3u8 failed.
> All candidate resources failed to load. Media load paused.
> All candidate resources failed to load. Media load paused.
> GET http://sirisian.com/randomfiles/testhls/hls/hdmi/index.m3u8
> GET http://sirisian.com/randomfiles/testhls/hls/hdmi/index.m3u8
> ... and it continues getting the index.m3u8 forever.

In the network tab it shows it request those files successfully.

In Chrome nothing happens and nothing is displayed in the console, but in the network tab I can see the following successful requests:

> index.html
> video.min.js
> videojs-contrib-hls.min.js
> index.m3u8
> 1460412184500.ts
> index.m3u8
> index.m3u8
> ... and the index.m3u8 repeats

**Comment by hdezela** [comment]
You're right, I get the same thing when disabling flash. That generally means something wrong with the encoding itself (flash is _very_ forgiving), what did you use to encode the original video and how did you then segment it?

Try playing around with different profiles and sizes, start at baseline 3.0 and step up from there.

This is what I use, which works with the setup described above:

`ffmpeg -y -async 1 -vsync -1 -analyzeduration 999999999 -i [INPUT]  -movflags faststart -keyint_min 30 -x264opts "keyint=30:min-keyint=30:no-scenecut" -g 30 -filter:v "scale=iw*min(1280/iw\,720/ih):ih*min(1280/iw\,720/ih), pad=1280:720:(1280-iw*min(1280/iw\,720/ih))/2:(720-ih*min(1280/iw\,720/ih))/2" -r:v 30 -s 1280x720 -b:v 1280k -c:v libx264 -pix_fmt yuv420p -profile:v baseline -level 3.1 -c:a libfdk_aac -ac 2 -ar 48000 -b:a 128k "720p.mp4"`

And for segmenting:

`ffmpeg -y -async 1 -vsync -1 -analyzeduration 999999999 -i "720p.mp4" -codec copy -map 0 -f segment -segment_list "720p.m3u8" -segment_time 10 -segment_list_type m3u8 -bsf:v h264_mp4toannexb "segment%05d.ts"`

**Comment by sirisian** [comment]
Currently I'm using:

ffmpeg -i http://192.168.1.168/hdmi -codec copy -bsf:a aac_adtstoasc -f flv rtmp://1270.0.0.1/live/hdmi

Maybe specifying the codec like you've done will help. I'll experiment a bit.

**Comment by hdezela** [comment]
Start super basic with something that's sure to be understood by even the most basic player like

`
ffmpeg -y -re -i http://192.168.1.168/hdmi -keyint_min 30 -x264opts "keyint=30:min-keyint=30:no-scenecut" -g 30 -r:v 30 -s 320x180 -b:v 256k -c:v libx264 -preset ultrafast -pix_fmt yuv420p -profile:v baseline -level 3.0 -c:a libfdk_aac -ac 1 -ar 48000 -b:a 96k -f flv rtmp://1270.0.0.1/live/hdmi
`

,and then work up (size,bitrate,profile) from there.

**Comment by sirisian** [comment]
woah, that works. Will continue to investigate. This could prove useful for others.

**Comment by matotiker** [comment]
I have the same issue .. it works on IE but there are issues on chrome and mozilla.
http://cdn.magictvbox.eu/eutv.m3u8 
it can be played on https://www.hlsplayer.net/

**Comment by sirisian** [comment]
Okay so I've tried multiple settings with my streaming device. constant/variable, changing the key frame interval, and the baseline or main profile. I also tried this at varying resolutions and bitrates and it happens everytime. Basically seems to be an ffmpeg bug or a problem with my stream. ffmpeg will run from anywhere between 10 second to minutes just fine then it starts throwing errors continuously and the stream dies.

> root@desktop:/tmp/hls/hdmi# ffmpeg -y -re -i http://192.168.1.168/hdmi -keyint_min 30 -x264opts "keyint=30:min-keyint=30:no-scenecut" -filter:v fps=25 -g 25 -r:v 30 -s 320x180 -b:v 256k -c:v libx264 -pix_fmt yuv420p -profile:v baseline -level 3.0 -c:a libfdk_aac -ac 2 -ar 48000 -b:a 96k -f flv rtmp://127.0.0.1:1935/live/hdmi;
> ffmpeg version N-79266-gcab9661 Copyright (c) 2000-2016 the FFmpeg developers
>   built with gcc 4.8 (Ubuntu/Linaro 4.8.1-10ubuntu9)
>   configuration: --prefix=/root/ffmpeg_build --pkg-config-flags=--static --extra-cflags=-I/root/ffmpeg_build/include --extra-ldflags=-L/root/ffmpeg_build/lib --bindir=/root/bin --enable-gpl --enable-libass --enable-libfdk-aac --enable-libx264 --enable-nonfree
>   libavutil      55. 20.100 / 55. 20.100
>   libavcodec     57. 34.100 / 57. 34.100
>   libavformat    57. 31.100 / 57. 31.100
>   libavdevice    57.  0.101 / 57.  0.101
>   libavfilter     6. 41.100 /  6. 41.100
>   libswscale      4.  1.100 /  4.  1.100
>   libswresample   2.  0.101 /  2.  0.101
>   libpostproc    54.  0.100 / 54.  0.100
> [h264 @ 0xa269f20] non-existing PPS 0 referenced
>     Last message repeated 1 times
> [h264 @ 0xa269f20] decode_slice_header error
> [h264 @ 0xa269f20] no frame!
> [h264 @ 0xa269f20] non-existing PPS 0 referenced
>     Last message repeated 1 times
> [snip, this happens for a while until it has a keyframe I believe. Not important]
> Input #0, mpegts, from 'http://192.168.1.168/hdmi':
>   Duration: N/A, start: 1096.192500, bitrate: N/A
>   Program 1 
>     Metadata:
>       service_name    : Service01
>       service_provider: FFmpeg
>     Stream #0:0[0x100]: Video: h264 (Baseline) ([27][0][0][0] / 0x001B), yuv420p, 1920x1080, 59.94 tbr, 90k tbn, 180k tbc
>     Stream #0:1[0x101]: Audio: aac (LC) ([15][0][0][0] / 0x000F), 44100 Hz, stereo, fltp, 152 kb/s
> [libx264 @ 0xa26e1a0] using cpu capabilities: MMX2 SSE2Fast SSSE3 SSE4.2
> [libx264 @ 0xa26e1a0] profile Constrained Baseline, level 3.0
> [libx264 @ 0xa26e1a0] 264 - core 148 - H.264/MPEG-4 AVC codec - Copyleft 2003-2016 - http://www.videolan.org/x264.html - options: cabac=0 ref=3 deblock=1:0:0 analyse=0x1:0x111 me=hex subme=7 psy=1 psy_rd=1.00:0.00 mixed_ref=1 me_range=16 chroma_me=1 trellis=1 8x8dct=0 cqm=0 deadzone=21,11 fast_pskip=1 chroma_qp_offset=-2 threads=6 lookahead_threads=1 sliced_threads=0 nr=0 decimate=1 interlaced=0 bluray_compat=0 constrained_intra=0 bframes=0 weightp=0 keyint=30 keyint_min=16 scenecut=0 intra_refresh=0 rc_lookahead=30 rc=abr mbtree=1 bitrate=256 ratetol=1.0 qcomp=0.60 qpmin=0 qpmax=69 qpstep=4 ip_ratio=1.40 aq=1:1.00
> Output #0, flv, to 'rtmp://127.0.0.1:1935/live/hdmi':
>   Metadata:
>     encoder         : Lavf57.31.100
>     Stream #0:0: Video: h264 (libx264) ([7][0][0][0] / 0x0007), yuv420p, 320x180, q=-1--1, 256 kb/s, 30 fps, 1k tbn, 30 tbc
>     Metadata:
>       encoder         : Lavc57.34.100 libx264
>     Side data:
>       cpb: bitrate max/min/avg: 0/0/256000 buffer size: 0 vbv_delay: -1
>     Stream #0:1: Audio: aac (libfdk_aac) ([10][0][0][0] / 0x000A), 48000 Hz, stereo, s16, 96 kb/s
>     Metadata:
>       encoder         : Lavc57.34.100 libfdk_aac
> Stream mapping:
>   Stream #0:0 -> #0:0 (h264 (native) -> h264 (libx264))
>   Stream #0:1 -> #0:1 (aac (native) -> aac (libfdk_aac))
> Press [q] to stop, [?] for help
> [h264 @ 0xa37ec20] negative number of zero coeffs at 0 0.53 bitrate= 281.5kbits/s speed=0.993x    
> [h264 @ 0xa37ec20] error while decoding MB 0 0
> [h264 @ 0xa37ec20] concealing 8160 DC, 8160 AC, 8160 MV errors in I frame
> [mpegts @ 0xa25c3a0] PES packet size mismatchme=00:00:10.03 bitrate= 283.1kbits/s speed=0.945x    
> [h264 @ 0xa34eda0] Increasing reorder buffer to 1
> [h264 @ 0xa34eda0] mb_type 45 in P slice too large at 2 8
> [h264 @ 0xa34eda0] error while decoding MB 2 8
> [h264 @ 0xa34eda0] concealing 7247 DC, 7247 AC, 7247 MV errors in P frame
> [h264 @ 0xa321f60] P sub_mb_type 19 out of range at 34 9
> [h264 @ 0xa321f60] error while decoding MB 34 9
> [h264 @ 0xa321f60] concealing 7095 DC, 7095 AC, 7095 MV errors in P frame
> [mpegts @ 0xa25c3a0] PES packet size mismatch
> [h264 @ 0xa351280] mb_type 46 in P slice too large at 113 9
> [h264 @ 0xa351280] error while decoding MB 113 9
> [h264 @ 0xa351280] concealing 7016 DC, 7016 AC, 7016 MV errors in P frame
> [h264 @ 0xa37ec20] mb_type 59 in P slice too large at 44 11
> [h264 @ 0xa37ec20] error while decoding MB 44 11
> [h264 @ 0xa37ec20] concealing 6845 DC, 6845 AC, 6845 MV errors in P frame
> [h264 @ 0xa3fcc00] mb_type 64 in P slice too large at 1 8
> [h264 @ 0xa3fcc00] error while decoding MB 1 8
> [h264 @ 0xa3fcc00] concealing 7248 DC, 7248 AC, 7248 MV errors in P frame
> [h264 @ 0xa34eda0] mb_type 88 in P slice too large at 45 9
> [h264 @ 0xa34eda0] error while decoding MB 45 9
> [h264 @ 0xa34eda0] concealing 7084 DC, 7084 AC, 7084 MV errors in P frame
> [h264 @ 0xa351280] out of range intra chroma pred mode at 61 8trate= 317.6kbits/s speed=0.888x    
> [h264 @ 0xa351280] error while decoding MB 61 8
> [h264 @ 0xa351280] concealing 7188 DC, 7188 AC, 7188 MV errors in P frame
> [h264 @ 0xa3fcc00] mb_type 140 in P slice too large at 1 8
> [h264 @ 0xa3fcc00] error while decoding MB 1 8
> [h264 @ 0xa3fcc00] concealing 7248 DC, 7248 AC, 7248 MV errors in P frame
> [h264 @ 0xa321f60] cbp too large (58) at 13 8me=00:00:11.14 bitrate= 292.0kbits/s speed=0.934x    
> [h264 @ 0xa321f60] error while decoding MB 13 8
> [h264 @ 0xa321f60] concealing 7236 DC, 7236 AC, 7236 MV errors in P frame
> [h264 @ 0xa37ec20] mb_type 40 in P slice too large at 2 8
> [h264 @ 0xa37ec20] error while decoding MB 2 8
> [h264 @ 0xa37ec20] concealing 7247 DC, 7247 AC, 7247 MV errors in P frame
> [mpegts @ 0xa25c3a0] PES packet size mismatchme=00:00:11.19 bitrate= 292.4kbits/s speed=0.899x    
> [h264 @ 0xa34eda0] mb_type 134 in P slice too large at 4 8
> [h264 @ 0xa34eda0] error while decoding MB 4 8
> [h264 @ 0xa34eda0] concealing 7245 DC, 7245 AC, 7245 MV errors in P frame
> [mpegts @ 0xa25c3a0] PES packet size mismatch
> [h264 @ 0xa351280] mb_type 57 in P slice too large at 14 8
> [h264 @ 0xa351280] error while decoding MB 14 8
> [h264 @ 0xa351280] concealing 7235 DC, 7235 AC, 7235 MV errors in P frame
> [h264 @ 0xa34eda0] cbp too large (119) at 21 8
> [h264 @ 0xa34eda0] error while decoding MB 21 8
> [h264 @ 0xa34eda0] concealing 7228 DC, 7228 AC, 7228 MV errors in P frame
> [h264 @ 0xa3fcc00] cbp too large (640) at 38 8
> [h264 @ 0xa3fcc00] error while decoding MB 38 8
> [h264 @ 0xa3fcc00] concealing 7211 DC, 7211 AC, 7211 MV errors in P frame
> [h264 @ 0xa351280] out of range intra chroma pred mode at 100 14ate= 293.7kbits/s speed=0.875x    
> [h264 @ 0xa351280] error while decoding MB 100 14
> [h264 @ 0xa351280] concealing 6429 DC, 6429 AC, 6429 MV errors in P frame
> [h264 @ 0xa321f60] P sub_mb_type 4 out of range at 14 8
> [h264 @ 0xa321f60] error while decoding MB 14 8
> [h264 @ 0xa321f60] concealing 7235 DC, 7235 AC, 7235 MV errors in P frame
> [h264 @ 0xa37ec20] mb_type 130 in P slice too large at 9 8
> [h264 @ 0xa37ec20] error while decoding MB 9 8
> [h264 @ 0xa37ec20] concealing 7240 DC, 7240 AC, 7240 MV errors in P frame

This continues forever and never recovers.

Most of the time when it starts the first warning/error is "PES packet size mismatch". I haven't been able to figure out how to fix this. I find it interesting how everything works for a few seconds or minutes and then it breaks. It's like it can't recover from some corruption that's occurring. Might have to take this over to the ffmpeg group.

**Comment by hdezela** [comment]
Can you post the output of

`ffprobe -loglevel 40 -i http://192.168.1.168/hdmi`

**Comment by sirisian** [comment]
Here's the ffprobe you requested: http://pastebin.com/sVzbZ9Un

I think I've figured out partly what's happening. If ffmpeg can't encode faster than it reads it'll fail due to some kind of circular buffer overflow. Here's a run at 30 fps 1280x720@30Hz re-encoding http://pastebin.com/Qt9iV2m3 . The device can only encode at around 25 fps due to performance reasons and fails after a few minutes never recovering.

With -preset ultrafast where it manages to encode at 30 fps it never crashes. The quality is bad though at 768 kbps with ultrafast enabled. Also after about 5 minutes in video.js it stops playing. It continues to request new m3u8 files but no ts file is ever requested. The new .ts files on the machine are being generated fine are all being cycled just fine and the m3u8 file seems fine, but it just glitches and stops requesting the .ts files. If I refresh it starts playing again for 3 minutes then locks up up for 10 seconds continuing to request .ts files and then starts playing again then locks up 30 seconds later then stops requesting .ts files for good again while continuing to request new m3u8 files forever never recovering. I refresh this process seems to repeat. Also I've watched it for a bit and it gets really laggy for a while before it stops requesting ts files. The buffering pops up and it'll play for a second then buffer for what seems like 100 ms then play and do this a few times then stop requesting .ts files.

Not sure if any of that helps. (I'll be submitting an ffmpeg bug with my data once their bug tracker is back online).

**Comment by hdezela** [comment]
It looks like there is an issue with your source stream, the message "non-existing PPS 0 referenced" means its looking for a keyframe for the PPS and SPS info, but it is never received.

I wouldn't recommend filing a bug with ffmpeg as it is actually working quite heroically in this is case, its amazing the poor thing is able to output _anything_ without keyframes.

I'd take a look at what is generating the http://192.168.1.168/hdmi origin (it being http makes me think there might be an sdp file somewhere you could reference) to make sure it's sending at least one keyframe every 1s or so.

**Comment by sirisian** [comment]
These are the settings I have to play with on the HTTP device:

![settingsstream](https://cloud.githubusercontent.com/assets/1320998/14803427/0a98a686-0b1f-11e6-9c2e-124f638849b8.png)

That seems to indicate it's creating a keyframe once a second.

Also to be clear that error only happens when the encoder can't keep up. The video.js issue I explained in the second large paragraph still occurs even when there are no errors with ffmpeg. (It's still running fine for me). That seems to be an issue with video.js since I can watch it play for a few minutes, stop playing, and refresh to repeat. Something in video.js seems to trigger it to stop requesting new ts files even when the m3u8 is requested and references new ones.

**Comment by sirisian** [comment]
Interesting. I plugged in a Matrox Maevex 5150 I have and used ffmpeg to passthrough the 1920x1080@30Hz stream from rtsp to rtmp then used nginx-rtmp to convert it to hls.

In video.js after a few minutes (between 3 minutes and up to 30 minutes) of playing it stops requesting the new ts files and just continues to request m3u8 files. It runs fine in VLC. Is there nothing built into video.js or videojs-contrib-hls to recover from whatever error is occurring internally?

Interestingly when I refreshed once in a great while it would output in Firefox:

> VIDEOJS: ERROR: (CODE:2 MEDIA_ERR_NETWORK) A network error caused the media download to fail part-way

Not sure what that's about since it's playing just fine in VLC. Then when I refresh it starts playing again fine.

Also when it fails I see this in the console:

> InvalidStateError: An attempt was made to use an object that is not, or is no longer, usable.

That repeats over and over. If I refresh it starts playing again file for a long time until it fails again. I might try to use the non-minified version of the code and see if it tells me the line it failed at.

**Comment by mina64** [comment]
Hi there,

we have an hls link that our CDN provider is Akamai. everything works perfect with VideoJS plugin just it doesn't have sound on Desktop browsers. Do you have any idea how to fix it. Here is link:

http://www.kanatelevision.com/en/kana-tv-live-stream

Thanks for any response!

Regards,

**Comment by gesinger** [comment]
@sirisian , sorry this ticket has gone a long time without much attention. And @hdezela , thanks for doing so much debugging on it!

We've made a lot of changes to videojs-contrib-hls in the meantime. @sirisian , would you be able to give the latest version (as of this comment, v4.1.1) a shot?

Thanks again.

And @mina64 , would you be able to open a different issue (to keep this thread from going on with different problems)? Thanks!

**Comment by gesinger** [comment]
Closing for now. If you are still experiencing this issue with the latest version of videojs-contrib-hls (as of this comment, v5.0.0), please re-open the issue and provide a reduced test case using [this template](http://jsbin.com/vokipos/8/edit?html,output). Thank you.

**Comment by thebeachtoday** [comment]
@gesinger --- after a lot of issues, I found a root cause with a very specific setup that might help others.

Functional Issue: same as this: cannot play on Chrome, FF; same errors.  Will play on IE, IOS

Setup:  Stream from Raspberry Pi [i.e. Raspivid] piped to FFMPEG out to FLV to NGINX RTMP Module.  That is then creating the m3u8 and ts files for VideoJS [latest release, basic settings]

Technical Root Cause: hls_fragment_size, hls_playlist.  Fragment size was too small [was 500ms]; Playlist was too short [was 10s].   Now using the NGINX RTMP defaults [i.e. I just commented out the values from the nginx.conf file], and it works fine on all browsers.  It is choppy, but it works and is a place to start now.

Bottom line: nginx rtmp module settings in nginx.conf can cause weird errors and allow the stream to play just fine on certain platforms and not others.  With so many moving parts, hopefully this gives someone an idea how they might troubleshoot their issue that isn't obvious from the server log files or client error messages.  

Thanks for all the help; your support and everyone else saves me a ton of time [and money]!
